File: src/algorithms/dynamic-programming/subsequence/max-subarray-kadane/sources/max-subarray-kadane_test.go
package main

import "testing"

func TestMaxSubarrayKadaneClassicInput(t *testing.T) {
	if maxSubarrayKadane([]int{-2, 1, -3, 4, -1, 2, 1, -5, 4}) != 6 {
		t.Errorf("classic kadane input should return 6")
	}
}

func TestMaxSubarrayKadaneSinglePositive(t *testing.T) {
	if maxSubarrayKadane([]int{1}) != 1 {
		t.Errorf("single positive element should return 1")
	}
}

func TestMaxSubarrayKadaneSingleNegative(t *testing.T) {
	if maxSubarrayKadane([]int{-1}) != -1 {
		t.Errorf("single negative element should return -1")
	}
}

func TestMaxSubarrayKadaneMostlyPositive(t *testing.T) {
	if maxSubarrayKadane([]int{5, 4, -1, 7, 8}) != 23 {
		t.Errorf("[5,4,-1,7,8] should return 23")
	}
}

func TestMaxSubarrayKadaneAllNegative(t *testing.T) {
	if maxSubarrayKadane([]int{-3, -2, -1}) != -1 {
		t.Errorf("all negative should return least negative")
	}
}

File: src/algorithms/dynamic-programming/subsequence/max-subarray-kadane/max_subarray_kadane_test.py
import importlib
import sys
import os

sys.path.insert(0, os.path.join(os.path.dirname(__file__), "sources"))
mod = importlib.import_module("max-subarray-kadane")
max_subarray_kadane = mod.max_subarray_kadane

assert max_subarray_kadane([-2, 1, -3, 4, -1, 2, 1, -5, 4]) == 6, "classic kadane input should return 6"
assert max_subarray_kadane([1]) == 1, "single positive element should return 1"
assert max_subarray_kadane([-1]) == -1, "single negative element should return -1"
assert max_subarray_kadane([5, 4, -1, 7, 8]) == 23, "all mostly positive should return 23"
assert max_subarray_kadane([-3, -2, -1]) == -1, "all negative should return least negative"

if __name__ == "__main__":
    print("All tests passed!")
